### DISCUSSION TITLE
Unable to increment value in list of object.

### DISCUSSION THREAD

**Discussion by harshitsinghai77** [deta/cloud-docs#222] [Help]
Hey,

I have a use case where I need to increment id such as
```python
user = {
    "...other stuff"
    "analytics": [
        {"id": increment(), "task":"task1"},
        {"id": increment(), "task":"task2"},
        {"id": increment(), "task":"task3"}
    ]
}
```

My code
```python
def create_new_task(self, key: str, task: dict) -> NemoSettings:
        if not task:
            return
        task = NemoTasks(**task)
        task = task.dict()
        task["id"] = async_db.util.increment()
        new_task = {
            "tasks": async_db.util.append(task),
        }
        async_db.update(new_task, key)
```

But I'm getting an error saying:
`TypeError: Object of type Increment is not JSON serializable`
Which makes sense.

On print I get this
```python
{'id': <deta.base.Util.Increment object at 0x7f899005ac40>, 'created_at': '2022-05-26 08:12:51.714123', 'task_description': 'My Second Task', 'duration': 3600, 'task_date': '2022-05-26 08:12:51.714125'}
```

What am I doing wrong here?

**Comment by jnsougata** [comment]
`async_db.util.increment()` this method besically returns `Increment` type which is not a JSON serializable data type. You should assign some value of type `str` or `int` or `float` or `list` or `None` to the key `id` to use it as json and use it with `update` method in async_db.

Note: Python a dict can contain custom types / objets to any key but JSON can't as it relies on generic data types. Unlike JSON python keys can be integers too.

**Reply by harshitsinghai77** [reply]
I don't follow. I understood what you're trying to say about Dict and Json. 

Can you give a example? How should I use it my code?

**Reply by harshitsinghai77** [reply]
I want something like 

```python
[
  {
    "created_at": "2022-05-27 12:32:05.220952",
    "duration": 3600,
    "id": 1, # this should increment
    "task_date": "2022-05-27 12:32:05.220954",
    "task_description": "My Second Task"
  }
]
```

**Reply by harshitsinghai77** [reply]
I think `Increment` works when it is used in the root. Doesn't work when it's nested. 
```python
task = {**task, "id": async_db.util.increment()}
new_task = {
    "tasks": async_db.util.append(task),
}
async_db.update(new_task, google_id)
```

**Reply by harshitsinghai77** [reply]
```python
        if updates:
            for attr, value in updates.items():
                if isinstance(value, Util.Trim):
                    payload["delete"].append(attr)
                elif isinstance(value, Util.Increment):
                    payload["increment"][attr] = value.val
                elif isinstance(value, Util.Append):
                    payload["append"][attr] = value.val
                elif isinstance(value, Util.Prepend):
                    payload["prepend"][attr] = value.val
                else:
                    payload["set"][attr] = value

```

Deta is checking on the items. So this will satisfy `elif isinstance(value, Util.Append)`, but it's not checking the `value.val` types.
You get what am saying?

**Reply by rohanshiva** [reply]
Hello, 

Can you please clarify your situation?

If I understand it correctly, you are working with similar situation as follows:

Initially
```json
{
  "key": "key",
  "tasks": [
    {
      "id": 1,
      "task": "wake up"
    },
    {
      "id": 2,
      "task": "eat"
    }
  ]
}
```

After update
```json
{
  "key": "key",
  "tasks": [
    {
      "id": 1,
      "task": "wake up"
    },
    {
      "id": 2,
      "task": "eat"
    },
    {
      "id": 3,
      "task": "sleep"
    }
  ]
}

```

As far as I understand, you are dealing with a similar situation and you are trying to create a new task in the function specified. 

If so, you are correct the SDK doesn't support nested increments. However, here is a simple snippet that might help you.

```py
def get_task(task, id):
    res = {**task}
    res["id"] = id 
    return res

def create_new_task(key, task):
    item = db.get(key)
    del item["key"]

    task = get_task(task, len(item["tasks"])+1)

    updates = {"tasks": db.util.append(task)}

    db.update(updates, key)

create_new_task(key, {"task": "sleep"})
```
  
If this is your situation, here is another schema that might be better. Instead of storing all the tasks in a list, why not create another base and link them using the parent item's key. 

For example, you can have a base called `dates` and store items like:
```json
{"key": "parent_key", "date":"Monday, June 10, 2022"}
```

Now, you can have another base called `tasks` and store items like:

```py
{"key":"task1", "task": "wake up", "created_at": 1653674846, "date": "parent_key"} # created_at can be a UNIX timestamp
{"key":"task2", "task": "sleep", "created_at": 1653674847, "date": "parent_key"} 
```
This way you can also order your tasks based on the `created_at` field.

**Reply by harshitsinghai77** [reply]
Hey @rohanshiva 

Yes, you understood my use case correctly. I see both of your approaches, but can we not have SDK support nested increments? 

This is a typical use case, NoSQL databases generally have a nested object inside a list, and these objects need increments (maybe to delete the object by id). I think when `elif isinstance(value, Util.Append): payload["append"][attr] = value.val` we can do an extra step and when this condition is satisfied we can again call the function with parameter, something like that. I hope you understand.

**Reply by harshitsinghai77** [reply]
Also, in your second approach, can you tell me how to reference it when querying the database to fetch all tasks by a `user_id`?

**Reply by rohanshiva** [reply]
@harshitsinghai77 

Thank you so much for the feedback, we will look into it. 

Regarding your question about querying to fetch all tasks by a `user_id`, you can do something like the following:

Assuming you have a base with these items:

![image](https://user-images.githubusercontent.com/20916697/171239920-22ce8a6b-1de7-4758-9c7c-67cfb0f974a6.png)

```
query = {"user_id": "rohan"}
res = db.fetch(query)
all_tasks = res.items

# fetch until last is 'None'
while res.last:
    res = db.fetch(last=res.last)
    all_tasks += res.items
```

**Reply by harshitsinghai77** [reply]
Thanks.
